core/src/avm2/array.rs

```diff
@@ -2,17 +2,308 @@
 
 use crate::avm2::value::Value;
 use gc_arena::Collect;
+use std::collections::BTreeMap;
 use std::{cmp::max, ops::RangeBounds};
 
+const MIN_SPARSE_LENGTH: usize = 32;
+
 /// The array storage portion of an array object.
 ///
 /// Array values may consist of either standard `Value`s or "holes": values
 /// which are not properties of the associated object and must be resolved in
 /// the prototype.
 #[derive(Clone, Collect)]
 #[collect(no_drop)]
-pub struct ArrayStorage<'gc> {
-    storage: Vec<Option<Value<'gc>>>,
+pub enum ArrayStorage<'gc> {
+    Dense(Vec<Option<Value<'gc>>>, usize),
+    Sparse(BTreeMap<usize, Value<'gc>>, usize),
+}
+
+struct ArrayStorageIterator<'a, 'gc> {
+    storage: &'a ArrayStorage<'gc>,
+    index: usize,
+    index_back: usize,
+}
+
+pub struct ArrayStorageMutableIterator<'a, 'gc> {
+    storage: &'a mut ArrayStorage<'gc>,
+    index: usize,
+    index_back: usize,
+}
+
+impl<'a, 'gc> Iterator for ArrayStorageIterator<'a, 'gc> {
+    type Item = Option<Value<'gc>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match &self.storage {
+            ArrayStorage::Dense(storage, ..) => {
+                if self.index >= self.index_back {
+                    None
+                } else {
+                    let value = storage[self.index];
+                    self.index += 1;
+                    Some(value)
+                }
+            }
+            ArrayStorage::Sparse(storage, _length) => {
+                if self.index >= self.index_back {
+                    None
+                } else {
+                    let value = storage.get(&self.index).cloned();
```

Comment:
Is the `.cloned` necessary here (and in some other cases)? `Value` implements `Copy`.